### PR TITLE
API Auto Registration Tap Docs

### PR DESCRIPTION
**Pull request by jrshaffe**
Adds documentation for API Auto Registration

### CODE REVIEW

File: api-auto-registration/about.hbs.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# API Auto Registration
+
+## <a id='overview'></a> Overview
+
+When users deploy a workload into a run cluster that exposes an API, they want that API to automatically show in TAP GUI without needing any other manual steps. API Auto Registration is an automated workflow that will can use a supply chain to create and manage a k8s Custom Resource (CR) of type APIDescriptor, a controller to reconcile the CR and update the API entity in TAP GUI to achieve automated API registration from workloads. You can also use API Auto Registration without supply chains by directly applying an APIDescriptor to the cluster.
```
Comment (shrutiyer) [inline api-auto-registration/about.hbs.md:5]:
```suggestion
When users deploy a workload that exposes an API, they want that API to automatically show in TAP GUI without needing any other manual steps. API Auto Registration is an automated workflow that will can use a supply chain to create and manage a k8s Custom Resource (CR) of type APIDescriptor, a controller to reconcile the CR and update the API entity in TAP GUI to achieve automated API registration from workloads. You can also use API Auto Registration without supply chains by directly applying an APIDescriptor to the cluster.
```

File: api-auto-registration/installation.hbs.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+## <a id='parameters'></a>Parameters
+API Auto Registration uses the following parameters in `tap-values.yaml`
+
+```yaml
+api_autoregistration:
+  tap_gui_url: <URL TO TAP GUI>
+  cluster_name: <CLUSTER NAME AS SHOWN IN LIFECYCLE>
+```
+
+>**Note:** The value of `cluster_name` is what you will see in the `lifecycle` property of your API Entitites
```
Comment (shrutiyer) [inline api-auto-registration/installation.hbs.md:18]:
```suggestion
>**Note:** The value of `cluster_name` is what you will see in the `lifecycle` property of your API Entities within TAP GUI 
```

File: api-auto-registration/installation.hbs.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
+# Install API Auto Registration
+
+This document describes how to install API Auto Registration from the Tanzu Application Platform package repository.
+
+>**Note:** Use the instructions on this page if you do not want to use the full or run profile to install packages.
+Only the full and run profiles includes API Auto Registsration.
```
Comment (shrutiyer) [inline api-auto-registration/installation.hbs.md:6]:
```suggestion
The "full" and "run" profiles installs API Auto Registsration by default.
```

File: api-auto-registration/installation.hbs.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+When using a mutli-cluster installtion using the run cluster profile or without a profile you will need to set the `api_autoregistration` parameters manually.
+
+You can locate the `tap_gui_url` by going to the cluster that is running TAP GUI and executing the following commands
+
+```console
+kubectl get secret tap-values -n tap-install -o jsonpath="{.data['tap-values\.yaml']}" | base64 -d | yq '.tap_gui.app_config.backend.baseUrl'  
+``` 
+
+## <a id='prereqs'></a>Prerequisites
```
Comment (shrutiyer) [inline api-auto-registration/installation.hbs.md:36]: This is not a focused comment but this Prerequisites section seems to kinda break the flow? I feel like it should go before the Parameters section and remove/move the "Located the TAP GUI url using the instructions mentioned above" sentence out?

File: api-auto-registration/installation.hbs.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+     NAME                             VERSION        RELEASED-AT
+     apis.apps.tanzu.vmware.com       0.1.0          2022-08-09 16:27:06 -0400 EDT
+    ```  
+
+2. Create a config file named `api-auto-registration-config.yaml`.
+
+    Add the following intry into your `api-auto-registration-config.yaml` file.
+    ```yaml
+    tap_gui_url: YOUR-TAP-GUI-URL
+    cluster_name: API-LIFECYCLE
```
Comment (shrutiyer) [inline api-auto-registration/installation.hbs.md:66]: We intentionally left lifecycle details out from this config. 
```suggestion
    cluster_name: YOUR-CLUSTER-NAME
```

File: api-auto-registration/usage.hbs.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+In order to auto register your api with tap gui you just need to make a couple of modifications to your workload yaml.
+
+1. Add the label `apis.apps.tanzu.vmware.com/register-api: "true"`
+1. Add a param of `type api_descriptor`
+
+```yaml
+  params:
+    - name: api_descriptor
+      value:
+        type: openapi   # One of openapi, aysnc, graphql, gRPC
```
Comment (shrutiyer) [inline api-auto-registration/usage.hbs.md:14]:
```suggestion
        type: openapi   # One of openapi, aysncapi, graphql, grpc
```

File: api-auto-registration/usage.hbs.md [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        description: "A set of API endpoints to manage the resources within the petclinic app."
+```
+
+There are 2 different options for the location. 
+
+The default supply chains use knative to deploy your applications. In this event the only location information you need to send is the path to the api documentation. The controller can figure out the base url for you.
+
+Another option is you can hard code the url using the baseURL property.  The controller will use a combination of this baseURL and your path to retrieve the yaml
+
+Example workload:
```
Comment (shrutiyer) [inline api-auto-registration/usage.hbs.md:27]:
```suggestion
Example workload that exposes a knative service:
```

File: api-auto-registration/usage.hbs.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+```yaml
+apiVersion: carto.run/v1alpha1
+kind: Workload
+metadata:
+  name: petclinic-knative-02
```
Comment (shrutiyer) [inline api-auto-registration/usage.hbs.md:33]:
```suggestion
  name: petclinic-knative
```

File: api-auto-registration/usage.hbs.md [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+  description: A set of API endpoints to manage the resources within the petclinic app.
+  system: spring-petclinic
+  owner: team-petclinic
+  location:
+    path: "/test/openapi"
+    baseURL:
+      ref:
+        apiVersion: projectcontour.io/v1
+        kind: HTTPProxy
+        name: wiremock
```
Comment (shrutiyer) [inline api-auto-registration/usage.hbs.md:121]: Let's remove wiremock reference. Something like so or anything else
```suggestion
        name: my-httpproxy
```

File: api-auto-registration/about.hbs.md (hunk 1 above)
Comment (LittleBaiBai) [inline api-auto-registration/about.hbs.md:5]:
```suggestion
When users deploy a workload that exposes an API, they want that API to automatically show in TAP GUI without needing any other manual steps. The ootb supply chain creates a k8s Custom Resource (CR) of type `APIDescriptor` for workloads that expressed the intention to expose API, and API Auto Registration is packages that reconciles the CR and register/manage the API entity in TAP GUI to achieve automated API registration from workloads. You can also use API Auto Registration without supply chains by directly applying an APIDescriptor to the run clusters.
```

File: api-auto-registration/installation.hbs.md [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+```yaml
+api_autoregistration:
+  tap_gui_url: <URL TO TAP GUI>
+  cluster_name: <CLUSTER NAME AS SHOWN IN LIFECYCLE>
+```
+
+>**Note:** The value of `cluster_name` is what you will see in the `lifecycle` property of your API Entitites 
+
+If you are installing with full profile you will get the following default values, but keep in mind you can override them if need be.
```
Comment (LittleBaiBai) [inline api-auto-registration/installation.hbs.md:20]:
```suggestion
If you are installing TAP with run or full profile you will get the following default values, but keep in mind you can override them if need be.
```

File: api-auto-registration/installation.hbs.md [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+```yaml
+api_autoregistration:
+  tap_gui_url: http://server.tap-gui.svc.cluster.local:7000 
+  cluster_name: <Value from shared.ingressDomain or 'dev' if not found>
+```
+
+When using a mutli-cluster installtion using the run cluster profile or without a profile you will need to set the `api_autoregistration` parameters manually.
```
Comment (LittleBaiBai) [inline api-auto-registration/installation.hbs.md:28]:
```suggestion
When using a mutli-cluster installtion using the run cluster profile or without a profile you will need to set the `api_auto_registration.tap_gui_url` parameters correctly for successful entity registration with TAP GUI, and you are strongly recommended to set `cluster_name` to a unique value for each run cluster.
```

File: api-auto-registration/installation.hbs.md [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+
+You can locate the `tap_gui_url` by going to the cluster that is running TAP GUI and executing the following commands
```
Comment (LittleBaiBai) [inline api-auto-registration/installation.hbs.md:30]:
```suggestion
You can locate the `tap_gui_url` by going to the view cluster with the TAP GUI you want to register the entity with and executing the following commands
```

File: api-auto-registration/installation.hbs.md (hunk 5 above)
Comment (LittleBaiBai) [reply api-auto-registration/installation.hbs.md:66]: I like the suggested change but also would like to add some explanation below on the usage of this field, and note that it should be unique for each run cluster.

File: api-auto-registration/installation.hbs.md [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    Add the following intry into your `api-auto-registration-config.yaml` file.
+    ```yaml
+    tap_gui_url: YOUR-TAP-GUI-URL
+    cluster_name: API-LIFECYCLE
+    ```
+
+3. Install the package using the Tanzu CLI
+   
+   ```console
+   tanzu package install api-autoregistration
```
Comment (LittleBaiBai) [inline api-auto-registration/installation.hbs.md:72]:
```suggestion
   tanzu package install api-auto-registration 
```

File: api-auto-registration/installation.hbs.md [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+>**Note:** Use the instructions on this page if you do not want to use the full or run profile to install packages.
+Only the full and run profiles includes API Auto Registsration.
+For more information about profiles, see [About Tanzu Application Platform components and profiles](../about-package-profiles.md).
+
+## <a id='parameters'></a>Parameters
+API Auto Registration uses the following parameters in `tap-values.yaml`
+
+```yaml
+api_autoregistration:
```
Comment (LittleBaiBai) [inline api-auto-registration/installation.hbs.md:13]:
```suggestion
api_auto_registration:
```

File: api-auto-registration/installation.hbs.md [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+  tap_gui_url: <URL TO TAP GUI>
+  cluster_name: <CLUSTER NAME AS SHOWN IN LIFECYCLE>
+```
+
+>**Note:** The value of `cluster_name` is what you will see in the `lifecycle` property of your API Entitites 
+
+If you are installing with full profile you will get the following default values, but keep in mind you can override them if need be.
+
+```yaml
+api_autoregistration:
```
Comment (LittleBaiBai) [inline api-auto-registration/installation.hbs.md:23]:
```suggestion
api_auto_registration:
```

File: api-auto-registration/usage.hbs.md [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+```yaml
+apiVersion: carto.run/v1alpha1
+kind: Workload
+metadata:
+  name: petclinic-knative-02
+  labels:
+    apps.tanzu.vmware.com/workload-type: web
+    apps.kubernetes.io/name: spring-petclinic
+    apps.tanzu.vmware.com/has-tests: "true"
+    app.kubernetes.io/part-of: spring-petclinic
```
Comment (LittleBaiBai) [inline api-auto-registration/usage.hbs.md:38]: We can link to the workload file in our app-accelerator template project for a full version users can try with. Maybe here we should only focus on the additional things our workflow cares?
```suggestion
    ...
```

File: api-auto-registration/usage.hbs.md [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+    apis.apps.tanzu.vmware.com/register-api: "true" 
+spec:
+  source:
+    git:
+      url: https://github.com/sample-accelerators/tanzu-java-web-app.git
+      ref:
+        branch: accelerator
```
Comment (LittleBaiBai) [inline api-auto-registration/usage.hbs.md:45]:
```suggestion
    ...
```

File: api-auto-registration/usage.hbs.md [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+        type: openapi
+        location: 
+          baseURL: http://petclinic-hard-coded.my-apps.tapdemo.vmware.com/    
+          path: "/v3/api-docs"
+        owner: team-petclinic
+        system: pet-clinics
+        description: "A set of API endpoints to manage the resources within the petclinic app."
+```
+
+## <a id='without-supply-chain'></a>Without Supply Chains
```
Comment (LittleBaiBai) [inline api-auto-registration/usage.hbs.md:82]:
```suggestion
## <a id='without-supply-chain'></a>Without the OOTB Supply Chains
```

File: api-auto-registration/usage.hbs.md [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+You can also utilize API Auto Registration if you are not using supply chains. You can do this by directly creating an APIDescriptor custom resource.
```
Comment (LittleBaiBai) [inline api-auto-registration/usage.hbs.md:83]: The information in this section is also useful for operators that writes their own custom supply chains. Maybe we can capture both use cases here?

**Review by LittleBaiBai — CHANGES_REQUESTED** [review]

File: api-auto-registration/installation.hbs.md [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+
+If you are installing TAP with run or full profile you will get the following default values, but keep in mind you can override them if need be.
+
+```yaml
+api_auto_registration:
+  tap_gui_url: http://server.tap-gui.svc.cluster.local:7000 
+  cluster_name: <Value from shared.ingressDomain or 'dev' if not found>
+```
+
+When using a mutli-cluster installtion using the run cluster profile or without a profile you will need to set the `api_auto_registration.tap_gui_url` parameters correctly for successful entity registration with TAP GUI, and you are strongly recommended to set `cluster_name` to a unique value for each run cluster.
```
Comment (visomar) [inline api-auto-registration/installation.hbs.md:28]:
```suggestion
When using a mutli-cluster installation using the run cluster profile or without a profile you will need to set the `api_auto_registration.tap_gui_url` parameters correctly for successful entity registration with TAP GUI, and you are strongly recommended to set `cluster_name` to a unique value for each run cluster.
```

File: api-auto-registration/installation.hbs.md [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+    Example output:
+
+    ```console
+     NAME                             VERSION        RELEASED-AT
+     apis.apps.tanzu.vmware.com       0.1.0          2022-08-09 16:27:06 -0400 EDT
+    ```  
+
+2. Create a config file named `api-auto-registration-config.yaml`.
+
+    Add the following intry into your `api-auto-registration-config.yaml` file.
```
Comment (visomar) [inline api-auto-registration/installation.hbs.md:63]:
```suggestion
    Add the following entry into your `api-auto-registration-config.yaml` file.
```

File: api-auto-registration/usage.hbs.md [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+          path: "/v3/api-docs"
+        owner: team-petclinic
+        system: pet-clinics
+        description: "A set of API endpoints to manage the resources within the petclinic app."
+```
+
+## <a id='without-supply-chain'></a>Without the OOTB Supply Chains
+You can also utilize API Auto Registration if you are not using supply chains. You can do this by directly creating an APIDescriptor custom resource.
+
+### <a id='prereqs'></a>With an Absolute URL
```
Comment (visomar) [inline api-auto-registration/usage.hbs.md:82]: Id's in the same page (specially the ones used with links) should be unique. These sections share the same id:

- Using Supply Chains
- With an Absolute URL
- With a Ref

File: api-auto-registration/usage.hbs.md (hunk 19 above)
Comment (jrshaffe) [reply api-auto-registration/usage.hbs.md]: Updated the wording to mention both scenarios

File: api-auto-registration/usage.hbs.md (hunk 22 above)
Comment (jrshaffe) [reply api-auto-registration/usage.hbs.md]: Gave them all unique IDs

File: api-auto-registration/installation.hbs.md (hunk 4 above)
Comment (jrshaffe) [reply api-auto-registration/installation.hbs.md:42]: Rearranged and removed the suggested sentence

**Review by BarrAndrea — APPROVED** [review]

**Pull request merged**